canary/producer-cpp/src/Canary.cpp

```diff
@@ -0,0 +1,899 @@
+    return STATUS_SUCCESS;
+}
+
+STATUS
+CanaryStreamCallbackProvider::streamErrorReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                       UPLOAD_HANDLE upload_handle, UINT64 errored_timecode, STATUS status_code) {
+    LOG_ERROR("Reporting stream error. Errored timecode: " << errored_timecode << " Status: "
+                                                           << status_code);
+    CustomData *data = reinterpret_cast<CustomData *>(custom_data);
+    bool terminate_pipeline = false;
+
+    if ((!IS_RETRIABLE_ERROR(status_code) && !IS_RECOVERABLE_ERROR(status_code))) {
+        data->stream_status = status_code;
+        terminate_pipeline = true;
+    }
+
+    if (terminate_pipeline && data->main_loop != NULL) {
+        LOG_WARN("Terminating pipeline due to unrecoverable stream error: " << status_code);
+        g_main_loop_quit(data->main_loop);
+    }
+
+    return STATUS_SUCCESS;
+}
+
+STATUS
+CanaryStreamCallbackProvider::droppedFrameReportHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                        UINT64 dropped_frame_timecode) {
+    LOG_WARN("Reporting dropped frame. Frame timecode " << dropped_frame_timecode);
+    return STATUS_SUCCESS;
+}
+
+STATUS
+CanaryStreamCallbackProvider::fragmentAckReceivedHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                         UPLOAD_HANDLE upload_handle, PFragmentAck pFragmentAck) {
+    CustomData *data = reinterpret_cast<CustomData *>(custom_data);
+    if (pFragmentAck->ackType != FRAGMENT_ACK_TYPE_PERSISTED && pFragmentAck->ackType != FRAGMENT_ACK_TYPE_RECEIVED)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    cout << "Map Debug: searching in map for a key of pFragmentAck->timestamp: " << pFragmentAck->timestamp << endl;
+    map<uint64_t, uint64_t>::iterator iter;
+    iter = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp);
+    if(iter == data->timeOfNextKeyFrame->end())
+    {
+        cout << "Map Debug: TimeOfNextKeyFrame key-value pair not present in map" << endl;
+    }
+    else
+    {
+        cout << "Map Debug: Found TimeOfNextKeyFrame key-value pair in map" << endl;
+        uint64_t timeOfFragmentEndSent = data->timeOfNextKeyFrame->find(pFragmentAck->timestamp)->second;
+        cout << "Map Debug: with value of " << timeOfFragmentEndSent << endl;
+        cout << "Map Debug: latency: " << duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count() - timeOfFragmentEndSent << endl;
+
+        // When Canary sleeps, timeOfFragmentEndSent become less than currentTimeStamp, don't send those metrics
+        if (timeOfFragmentEndSent > pFragmentAck->timestamp)
+        {
+            if (pFragmentAck->ackType == FRAGMENT_ACK_TYPE_PERSISTED)
+            {
+                Aws::CloudWatch::Model::MetricDatum persistedAckLatency_datum;
```

Comment:
General guidance: Let us following one style of naming: either camel case: `persistedAckLatencyDatum` or `persisted_ack_latency_datum`. Let us mix it. Since most of the code is using former, please modify relevant function names and variable names to be all camel cases type.

Reply:
Fixed this for everything but the iot credential and canary callback stuff - seems to make more sense to leave as is for these two.